### DISCUSSION TITLE
Playwright traces now supported in Artillery Cloud

### DISCUSSION THREAD

**Discussion by hassy** [artilleryio/artillery#3221] [Announcements]
Support for Playwright tracing is now generally available in Artillery Cloud.

![Playwright Traces in Artillery](https://github.com/artilleryio/artillery/assets/1490/047022e5-d9af-41b9-b72e-a224f8d36ff9)

## What is this?

Artillery can now automatically record and upload [Playwright traces](https://playwright.dev/docs/trace-viewer) for failing VUs in a Playwright load test.

Recorded traces are available under the **Traces** tab in a load test report and can be viewed inline with Playwright Trace Viewer.

Playwright traces can be recorded for any tests - running locally with `artillery run`, or on AWS ECS with `artillery run-fargate`.

## Who is this for?

This feature is for anyone running load tests with Playwright and Artillery. Playwright trace recordings make it easier to debug and fix issues that cause VUs to fail.

## How do I get started?

Enable Playwright trace recording in your Artillery config:

```yaml
config:
  engines:
    playwright:
      trace:
        enabled: true
```

Run your Artillery test as normal, with reporting to Artillery Cloud turned on:

```sh
artillery run-fargate my-test.yml --record --key $ARTILLERY_CLOUD_KEY
```
Get your API key via **Settings** -> **API Keys** in Artillery Cloud (https://app.artillery.io)

## How much does this cost?

This feature is available to all users of Artillery Cloud. For users on the free tier, up to 5 traces can be recorded per test, with trace retention period of 7 days.

## Demo

https://github.com/artilleryio/artillery/assets/1490/c6db9ccd-aaa1-4945-a254-bbaac2a1da48

**Comment by Rio517** [comment]
Is there any way to capture the traces locally? I have an issue where some third-party asset is causing timeouts, and I am trying to run the tests locally against localhost with network access disabled. 

For now, I've been doing this, essentially.

  ```js
  const randomID = Math.random().toString(36).slice(2);
  const tracePath = `traces/trace-${randomID}.zip`;
  const context = page.context();
  await context.tracing.start({ screenshots: true, snapshots: true });
  try {
    //code that may fail or the whole test function.
  } finally {
    console.log('randomID', randomID, 'Error - Trace Captured')
    await context.tracing.stop({ path: tracePath });
  }
  ```

**Reply by hassy** [reply]
You can set the `PLAYWRIGHT_TRACING_OUTPUT_DIR` environment variable to tell Artillery where to save the traces ([relevant code here](https://github.com/artilleryio/artillery/blob/main/packages/artillery-engine-playwright/index.js#L87)). It's an undocumented option but won't be going away.

**Reply by Rio517** [reply]
Awesome. Thanks!
